llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -15267,6 +15267,48 @@ static SDValue reverseZExtICmpCombine(SDNode *N, SelectionDAG &DAG,
   return DAG.getNode(ISD::ZERO_EXTEND, DL, VT, Res);
 }
 
+static SDValue reduceANDOfAtomicLoad(SDNode *N,
+                                     TargetLowering::DAGCombinerInfo &DCI) {
+  SelectionDAG &DAG = DCI.DAG;
+  if (N->getOpcode() != ISD::AND)
+    return SDValue();
+
+  SDValue N0 = N->getOperand(0);
+  if (N0.getOpcode() != ISD::ATOMIC_LOAD)
+    return SDValue();
+
+  AtomicSDNode *ALoad = cast<AtomicSDNode>(N0.getNode());
+  if (isStrongerThanMonotonic(ALoad->getSuccessOrdering()))
+    return SDValue();
+
+  EVT LoadedVT = ALoad->getMemoryVT();
+  EVT ResultVT = N->getValueType(0);
+
+  SDValue MaskVal = N->getOperand(1);
+  ConstantSDNode *MaskConst = dyn_cast<ConstantSDNode>(MaskVal);
+  if (!MaskConst)
+    return SDValue();
+  uint64_t Mask = MaskConst->getZExtValue();
+  uint64_t ExpectedMask = LoadedVT.getSizeInBits() == 8    ? 0xFF
+                          : LoadedVT.getSizeInBits() == 16 ? 0xFFFF
+                          : LoadedVT.getSizeInBits() == 32 ? 0xFFFFFFFF
+                                                           : 0xFFFFFFFFFFFFFFFF;
+  if (Mask != ExpectedMask)
+    return SDValue();
+
+  SDLoc DL(N);
+  SDValue Chain = ALoad->getChain();
+  SDValue Ptr = ALoad->getBasePtr();
+  MachineMemOperand *MemOp = ALoad->getMemOperand();
+  SDValue ZextLoad = DAG.getExtLoad(ISD::ZEXTLOAD, DL, ResultVT, Chain, Ptr,
+                                    MemOp->getPointerInfo(), LoadedVT,
+                                    MemOp->getAlign(), MemOp->getFlags());
+  DAG.ReplaceAllUsesOfValueWith(SDValue(N, 0), ZextLoad);
```

Comment:
We should use `DCI.CombineTo(N, ZextLoad)` instead of `DAG.ReplaceAllUsesOfValueWith`